### ISSUE TITLE
TF 2.2 - Train_step output control

### ISSUE THREAD

**Issue by ManiadisG** [tensorflow/tensorflow#39121]
**System information**
- TensorFlow version (you are using): 2.2 rc1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
TF 2.2 allows modifying train_step() and make_train_function() to leverage the advantages of fit() with a lot of customization. However, due to train_step() being a tf.function, there is no way to extract variables from it (e.g. the model's output during training), other than by adding it to the *outputs* dictionary which goes directly to the logs. By customizing make_train_function() it is possible to do that even with multiple GPUs (concatenating accross devices), but adding that output to the logs creates a domino of problems when using callbacks.

Below is an example of mine where the model's outputs are stored with key *output* in the dict, concatenated accross devices, and returned to the dict, which is stored to the logs during each step in fit().

```
# In make_train_function()
...
def train_function(iterator):
    data = next(iterator)
    output_dict = self.distribute_strategy.run(
        self.train_step, args=(data,))
    m_output = output_dict.pop('output')
    m_output = reduce_per_replica(
        m_output, self.distribute_strategy, reduction='concat')
    output_dict = reduce_per_replica(
        output_dict, self.distribute_strategy, reduction='first')
    output_dict['output'] = m_output
    return output_dict
...

# Then in fit (without changes)
...
tmp_logs = train_function(iterator)
if not data_handler.inferred_steps:
  context.async_wait()
logs = tmp_logs
...
```

My suggestion would be changing the fit() function so that instead of assigning the output of train_function() to *tmp_logs*, it allows for additional outputs from train_step(), stored in a distinct variable which can then be accessed via callbacks. Below I have modified train_function() and fit() to return a list that would allow storing desired outputs in a separate variable.

```
def train_function(iterator):
    data = next(iterator)
    output_dict = self.distribute_strategy.run(
        self.train_step, args=(data,))
    output_dict = reduce_per_replica(
        output_dict, self.distribute_strategy, reduction='first')
    return output_dict, None

#OR with user modifications to get some output

def train_function(iterator):
    data = next(iterator)
    output_dict = self.distribute_strategy.run(
        self.train_step, args=(data,))
    m_output = output_dict.pop('output')
    m_output = reduce_per_replica(
        m_output, self.distribute_strategy, reduction='concat')
    output_dict = reduce_per_replica(
        output_dict, self.distribute_strategy, reduction='first')
    return output_dict, m_output

# Then in fit
...
tmp_logs, tmp_step_output = train_function(iterator)
if not data_handler.inferred_steps:
  context.async_wait()
logs, step_output = tmp_logs, tmp_step_output
...
```
With this modification and by giving all callbacks access to *step_output* users can customise train_function() and get whatever they want outside the training graph. These modifications would give them absolute freedom over it.

**Will this change the current api? How?**
As far as I can see the only changes would be the above code snippets in fit() and make_train_function(), and then adding *step_output* to the default inputs for callbacks.

**Who will benefit with this feature?**
Everyone. Other than model outputs it can be used to store gradients, and intermediate outcomes. It would allow for much better prototyping with minimal changes to the API.

**Any Other info.**

**Comment by omalleyt12** [comment]
@ManiadisG Thanks for the issue!

> but adding that output to the logs creates a domino of problems when using callbacks.

Could you provide example of the errors you've seen? It's probably best to fix the problem at this level IMO

**Comment by ManiadisG** [comment]
The core of the issue is that the logs variable is handled in a specific way by standard Callbacks that only makes sense for losses and metrics. In my case I wanted to extract the model's output during training so the problems I encountered are based on that. The output was a (N, X) matrix where N is the batch size, and I stored it to the logs dict with a "output" key.

I identified the following issues.
1. The Progbar callback printed the "output" as a value (same as the error), I'm assuming via reduce_mean. After an epoch was over it tried to aggregate the result which led to errors if the batches were uneven (due to the last batch not necessarily being the same size as the others).
2. The Tensorboard callback registered "output" and presented it in the report same as the losses, again I assume via reduce_mean. I did not test if this leads to errors or if it is just an aesthetic issue, but I'm assuming it would also cause performance issues if I tried to extract larger data and Tensorboard kept track of them.
3. The History callback also registered "output" same as a loss and again sometimes produced errors in the aggregation.

Some of those issues could be solved by using almost exclusively customized Callbacks, which I tried to do with some success. I also tried a Callback that pops "output" from the logs dictionary but it appears that other Callbacks access the logs first so it didn't really work. I also couldn't solve possible performance issues from the model possibly having to process/store the output accross an epoch's steps. As of 2.2-rc1, CallbackList is very hard to configure outside of fit() and there is very limited user visibility and control over what the model does with logs by default.

I would like to point out though that, even if that was not the case, having to rewrite every Callback negates a lot of the advantages of using fit(). And even if these errors were solved individually, not letting the users extract data from the training step without having pre-determined operations applied seems very error-prone.

To solve this at the Callback level, may I suggest defining a specific key (e.g. "user_output") that all standard Callbacks ignore. That way users would be able to store whatever they want in the logs dictionary under a specific key and handle it in any way they see fit with their own custom Callbacks.

Alternatively, adding an exception argument to the Callback class that can be passed via fit() to CallbackList would also work. That could take the form of a list of keys (defaulting to None), so that standard Callbacks will ingore the corresponding logs.

**Comment by ManiadisG** [comment]
Update.

It turns out there is only an error when using the Progbar callback, which at the end of the epoch tries to aggregate the logs. The History and Tensorboard Callbacks do not cause errors and can be overwritten to avoid performance and aesthetic problems.

Basically it appears that, in my case, the issue can be overcome by customizing Callbacks. Progbar in particular needs to be overwritten and the CallbackList object must be initialized outside fit with tensorflow.python.keras.callbacks.configure_callbacks(). However I'm not sure what would happen with other kinds of output data.

Regardless, since users can now customize train_step() and make_train_step(), it seems that there should be a way to add values to the train step function's output that are are not treated as metrics by standard Callbacks. It is the best way to guarantee that errors are avoided without restricting the output type and having to rewrite multiple callbacks.

**Comment by omalleyt12** [comment]
@ManiadisG Thanks for the update!

I think we should definitely fix `ProgbarLogger`, `History`, and `TensorBoard` so that they only try to handle log items that are scalars

Re adding another output to `Model.train_step`, I'm wary of adding another return value to the function. One way I've seen people handle this kind of use case is to smuggle the value out by assigning it to a `tf.Variable`, something like:

```
import numpy as np
import tensorflow as tf

BATCH_SIZE = 2

class MyModel(tf.keras.Model):

  def __init__(self):
    super(MyModel, self).__init__()
    self.outputs = tf.Variable(tf.zeros((BATCH_SIZE, 10)),
                               trainable=False)
    self.layer = tf.keras.layers.Dense(10)

  def call(self, inputs):
    return self.layer(inputs)

  def train_step(self, data):
    x, y = data
    with tf.GradientTape() as tape:
      y_pred = self(x)
      self.outputs.assign(y_pred)
      loss = self.compiled_loss(
          y, y_pred, regularization_losses=self.losses)
    trainable_variables = self.trainable_variables
    gradients = tape.gradient(loss, trainable_variables)
    self.optimizer.apply_gradients(zip(gradients, trainable_variables))
    self.compiled_metrics.update_state(y, y_pred)
    return {m.name: m.result() for m in self.metrics}

class MyCallback(tf.keras.callbacks.Callback):
  def on_train_batch_end(self, batch, logs=None):
    print(tf.reduce_sum(self.model.outputs))

model = MyModel()
model.compile('sgd', 'mse')
x, y = np.ones((10, 100)), 10 * np.ones((10, 10))
model.fit(x, y, batch_size=BATCH_SIZE, callbacks=[MyCallback()], verbose=2)
```

Do you think that pattern would work for you? I still think medium-term we should fix the built-in Callbacks to handle this use case

**Comment by ManiadisG** [comment]
Thank you for your response and for going through the trouble to propose a solution.

I have solved my problem by overwritting all the Callbacks I needed. The only one that I couldn't directly overwrite and use by feeding it to fit() was Progbar. That was solved by creating the CallbackList object outside model.fit() with tensorflow.python.keras.callbacks.configure_callbacks(). BaseLogger, which looks like it would be the hardest to overwrite, doesn't look like it's causing problems.

Overall, the issue can be overcome with enough callback customization until a more concrete solution is added. However, I would point out that restricting Callbacks to scalar logs is still too broad. Most Callbacks already treat logs differently based on whether the key of the dict is in the stateful_metrics list. I think it would be cleaner to expand this function to allow users to pass exceptions to fit() or to CallbackList (when it is fully integrated to the API) so that the corresponding logs are ignored by default Callbacks.

WIth regard to the solution you proposed, I believe it it still has a weak point in that the use of tf.Variable and tf.assign assumes a fixed shape which is not always the case (an epoch's last batch may be smaller).

In any case, unless you would like the issue to remain open for future refference I believe that it can be closed.

**Comment by ManiadisG** [comment]
I solved the issue at the Callback level as you suggested. I tested it in various scenarios with the 2.2 version and it worked perfectly. I wrote a Colab notebook with the code implementing a classifier which you can find here:

https://github.com/ManiadisG/Customising-Tensorflow-2.2-Models

Other than a solution to this specific problem, in the notebook I also wrote custom save/load functions for Model subclasses that, in my opinion, are more flexible and reliable than the defaults, particularly with Model subclassing.
